### ISSUE TITLE
question about computation of the norm stats

### ISSUE THREAD

**Issue by yanan1116** [Physical-Intelligence/openpi#773]
hi,

I see that executing `uv run scripts/compute_norm_stats.py` is a must when fine-tuning on libero dataset.

however, when i fine-tune on droid dataset, and if i skip the compute norm stats step, it does not return any error.

may i ask, is it a must or recommendation when fine-tune on droid dataset where the my customized dataset is of joint position action format.

thanks.

here is my config for training
```python

    TrainConfig(

        name="pi05_droid___genesis_joint_position_2000_20fps",
        model=pi0_config.Pi0Config(
            pi05=True,
            action_dim=32,  # pi05 is trained with 32-dim actions
            action_horizon=16,
        ),
        data=LeRobotDROIDDataConfig(
            # Replace with your custom DROID LeRobot dataset repo id.
            repo_id="kaveh-kamali/genesis_joint_position_2000_20fps",
            base_config=DataConfig(prompt_from_task=True),
            assets=AssetsConfig(
                # Important: reuse the original DROID norm stats during fine-tuning!
                assets_dir="gs://openpi-assets/checkpoints/pi05_droid/assets",
                asset_id="droid",
            ),
        ),
        weight_loader=weight_loaders.CheckpointWeightLoader("gs://openpi-assets/checkpoints/pi05_droid/params"),
        num_train_steps=100_000,
        batch_size=32,
        #lora 
        freeze_filter=pi0_config.Pi0Config(
            action_dim=8, action_horizon=10, max_token_len=180, paligemma_variant="gemma_2b_lora"
        ).get_freeze_filter(),
        ema_decay=0.999,
        
        optimizer=_optimizer.AdamW(clip_gradient_norm=1.0),
        lr_schedule=_optimizer.CosineDecaySchedule(
            warmup_steps=10_000,
            peak_lr=5e-5,
            decay_steps=1_000_000,
            decay_lr=5e-5,
        ),
        wandb_enabled=False,
    ),

```

if i replace `pi05_droid` with `pi05_base`, it still does not return any error without running ` scripts/compute_norm_stats.py`

**Comment by yanan1116** [comment]
my customized dataset where the trajectories come from simulation env (Genesis).

not sure if i should compute the stats before fine-tuning.

**Comment by yanan1116** [comment]
i suppose that if not computing the norm stats, then the stats come from the original droid dataset ? which could be far away from my customized fine-tuning set ? which is not appropriate, right ?

**Comment by yanan1116** [comment]
here are suggestions from chatgpt, not sure it is right 



<html>
<body>
<html><head></head><body><p>Excellent question — this shows you understand one of the most subtle (and crucial) details of <strong>fine-tuning vision-language-action (VLA) models like Pi 0.5</strong>.</p>
<p>Let’s unpack it carefully 👇</p>
<hr>
<h2>🧩 Short answer</h2>
<p>✅ <strong>Yes — you <em>do</em> need to compute normalization (“norm”) statistics</strong> for your own trajectories <strong>if</strong> you are fine-tuning on <em>new simulator data (e.g., Genesis)</em> that likely has <strong>different scales or distributions</strong> (e.g. different joint ranges, action magnitudes, camera intrinsics, etc.)</p>
<p>But if your simulator’s data distribution matches the original Open-PI preprocessing (same units, action scaling, etc.), then you can reuse the pretrained stats.</p>
<hr>
<h2>🧠 Why this matters</h2>
<p>Pi 0.5 (and similar VLA models like RT-1, RT-X, or LeRobot-based models) expect each trajectory to be <strong>normalized</strong> before training:</p>
<p>[<br>
x_{norm} = (x - \mu) / \sigma<br>
]</p>
<p>where:</p>
<ul>
<li>
<p>( x ) is a raw observation (like end-effector position, gripper state, etc.)</p>
</li>
<li>
<p>( \mu ), ( \sigma ) are dataset-wide mean and std computed from <em>training data</em></p>
</li>
</ul>
<p>This is done for:</p>
<ul>
<li>
<p><strong>joint positions</strong></p>
</li>
<li>
<p><strong>actions</strong></p>
</li>
<li>
<p>sometimes <strong>images (pixel mean/std)</strong> or <strong>EE pose</strong></p>
</li>
</ul>
<hr>
<h2>🔍 In the Open-PI setup</h2>
<p>The checkpoint at</p>
<pre><code>gs://openpi-assets/checkpoints/pi05_base/params
</code></pre>
<p>was trained on their <em>internal</em> datasets (Real-World &amp; RTX-style).<br>
The model was normalized using <strong>their statistics</strong>, stored typically in files like:</p>
<pre><code>norm_stats.json
</code></pre>
<p>or</p>
<pre><code>dataset_stats.pkl
</code></pre>
<p>If your Genesis trajectories differ — for example, if your arm operates in different units (meters vs. cm) or camera normalization is different — then using their stats will mis-scale your data and <strong>harm fine-tuning</strong>.</p>
<hr>
<h2>🧮 So you should:</h2>
<ol>
<li>
<p><strong>Collect your dataset</strong> (N trajectories, each with observations/actions).</p>
</li>
<li>
<p><strong>Compute your own normalization stats</strong>, for example:</p>
<pre><code class="language-python">import numpy as np

all_actions = np.concatenate([traj["actions"] for traj in dataset], axis=0)
all_obs = np.concatenate([traj["observations"] for traj in dataset], axis=0)

norm_stats = {
    "actions": {
        "mean": all_actions.mean(axis=0).tolist(),
        "std": all_actions.std(axis=0).tolist(),
    },
    "observations": {
        "mean": all_obs.mean(axis=0).tolist(),
        "std": all_obs.std(axis=0).tolist(),
    },
}
</code></pre>
</li>
<li>
<p>Save as <code inline="">norm_stats.json</code> (or whatever format Pi 0.5 trainer expects).</p>
</li>
<li>
<p><strong>Pass that file path</strong> into your fine-tuning config or dataloader initialization.</p>
</li>
</ol>
<hr>
<h2>⚙️ Pro tip</h2>
<ul>
<li>
<p>Many Pi / Open-X models have code like:</p>
<pre><code class="language-python">from pi.models.utils import load_norm_stats
norm_stats = load_norm_stats(cfg.norm_path)
</code></pre>
<p>so if you supply your computed stats, they’ll override the pretrained ones automatically.</p>
</li>
<li>
<p>For small datasets (e.g. &lt;1k episodes), it can also help to do a <strong>“partial adaptation”</strong>:</p>
<ul>
<li>
<p>Start from pretrained stats,</p>
</li>
<li>
<p>Blend them with your new dataset’s stats (e.g. 80/20 weighting) to avoid overfitting to simulator noise.</p>
</li>
</ul>
</li>
</ul>
<hr>
<h2>✅ Summary</h2>

Case | Should you recompute norm stats? | Why
-- | -- | --
Using exact same dataset as base | ❌ No | Already normalized
Using similar data (same units, scaling) | Optional | Reuse pretrained stats
Using new simulator data (Genesis) | ✅ Yes | Different numeric scales
Using mixed datasets | ✅ Recommended | To unify distributions


<hr>
<p>Would you like me to show you an exact <strong>Python snippet</strong> that computes Pi 0.5-style norm stats (in the same format it expects, compatible with <code inline="">params</code> checkpoint)?</p></body></html>
</body>
</html>

**Comment by yanan1116** [comment]
<html>
<body>
<html><head></head><body><p>Excellent — that’s a really sharp follow-up 👏</p>
<p>So you’re now referring to <strong><code inline="">pi05_droid</code></strong>, not <code inline="">pi05_base</code>.<br>
That distinction <em>really matters</em> because of what the <strong>DROID</strong> dataset is and how it’s preprocessed.</p>
<p>Let’s go step-by-step 👇</p>
<hr>
<h2>🧩 Context: What <code inline="">pi05_droid</code> means</h2>
<p><code inline="">pi05_droid</code> is the variant of the Pi 0.5 model fine-tuned by the Open-PI team on the <strong>DROID dataset</strong>, which is:</p>
<ul>
<li>
<p>a <strong>massive real-robot dataset</strong> (from RT-X / Open X-Embodiment)</p>
</li>
<li>
<p>containing <strong>tens of thousands of trajectories</strong></p>
</li>
<li>
<p>all normalized and scaled consistently (via the <code inline="">rlds</code> format)</p>
</li>
</ul>
<p>So <code inline="">pi05_droid</code> isn’t trained from scratch — it’s <strong><code inline="">pi05_base</code> + DROID fine-tuning</strong>.</p>
<hr>
<h2>🧠 Key implication</h2>
<p>When you fine-tune from <code inline="">pi05_droid</code>, the model’s internal normalization assumptions (mean/std for actions, joints, gripper, etc.) are <strong>based on DROID statistics</strong>, <em>not on simulator data like Genesis</em>.</p>
<p>That means:</p>
<blockquote>
<p>✅ You should still compute your own normalization statistics<br>
if your Genesis trajectories’ scales differ from DROID’s.</p>
</blockquote>
<hr>
<h2>🧩 Why that matters (numerical example)</h2>
<p>In DROID:</p>
<ul>
<li>
<p>joint positions and end-effector positions are normalized assuming <strong>meters</strong>, roughly in <code inline="">[-1, 1]</code></p>
</li>
<li>
<p>actions are normalized torques or deltas, typically with <code inline="">std ≈ 0.1</code></p>
</li>
</ul>
<p>In Genesis or other simulators:</p>
<ul>
<li>
<p>end-effector positions might be in <strong>centimeters</strong></p>
</li>
<li>
<p>gripper actions might range <code inline="">[0, 255]</code> or <code inline="">[0, 1]</code></p>
</li>
<li>
<p>joint ranges may differ depending on URDF</p>
</li>
</ul>
<p>If you directly fine-tune without recomputing stats:</p>
<ul>
<li>
<p>the model may interpret a 10 cm movement as 10 m</p>
</li>
<li>
<p>gradients and policy updates blow up → poor convergence or NaNs</p>
</li>
</ul>
<hr>
<h2>🧮 What to do</h2>
<ol>
<li>
<p><strong>Compute your own normalization statistics</strong> from Genesis trajectories:</p>
<pre><code class="language-python">import numpy as np
import json

all_obs = np.concatenate([traj["observation/state"] for traj in trajectories], axis=0)
all_actions = np.concatenate([traj["action"] for traj in trajectories], axis=0)

stats = {
    "observation": {"mean": all_obs.mean(0).tolist(), "std": all_obs.std(0).tolist()},
    "action": {"mean": all_actions.mean(0).tolist(), "std": all_actions.std(0).tolist()},
}

with open("norm_stats.json", "w") as f:
    json.dump(stats, f, indent=2)
</code></pre>
</li>
<li>
<p><strong>Use that file</strong> in your fine-tuning config (e.g., via <code inline="">--norm-path norm_stats.json</code>).</p>
</li>
<li>
<p>Optionally <strong>initialize</strong> your stats from DROID’s and <em>blend</em> them:</p>
<pre><code class="language-python"># mix 80% DROID, 20% new dataset
mean = 0.8 * droid_mean + 0.2 * new_mean
std  = np.sqrt(0.8 * droid_std**2 + 0.2 * new_std**2)
</code></pre>
<p>This can help stability if your dataset is small.</p>
</li>
</ol>
<hr>
<h2>✅ Summary</h2>

Base Checkpoint | Dataset Source | Need to Recompute Norm Stats? | Why
-- | -- | -- | --
pi05_base | Internal Open-PI mix | ✅ Yes | Your simulator likely differs
pi05_droid | Real-world DROID (rlds) | ✅ Yes (recommended) | Genesis data has different units/scales
Your own continued training on DROID data | ❌ No | Already consistent |  


<hr>
<h2>🧩 Bonus tip</h2>
<p>When you run fine-tuning, always visualize the normalized actions:</p>
<pre><code class="language-python">import matplotlib.pyplot as plt
plt.hist(normed_actions.flatten(), bins=100)
</code></pre>
<p>You want them roughly centered near 0 with σ≈1.<br>
If they’re huge or tiny, your stats are off.</p>
<hr>
<p>Would you like me to show you how to <strong>extract the original DROID norm stats</strong> (so you can compare them numerically to your Genesis stats before fine-tuning)?</p></body></html>
</body>
</html>
